tests/test_layabout.py

```diff
@@ -1 +1,327 @@
+    )
+
+    # Make sure the SlackClient was only instantiated once so we know that we
+    # reused the existing instance.
+    SlackClient.assert_called_once_with(token=TOKEN)
+
+
+def test_layabout_can_continue_after_successful_reconnection(monkeypatch):
+    """
+    Test that layabout can continue to handle events after successfully
+    reconnecting to the Slack API.
+    """
+    layabout = Layabout(name='test')
+
+    connections = [
+        True,  # Succeed with the first connection (_connect).
+        True,  # Succeed later with a reconnection (_reconnect).
+    ]
+    reads = [
+        TimeoutError,  # Raise an exception on the first read.
+        [],  # Return empty events on the second read (after reconnection).
+    ]
+
+    rtm_connect = MagicMock(side_effect=connections)
+    rtm_read = MagicMock(side_effect=reads)
+    slack = MagicMock(rtm_connect=rtm_connect, rtm_read=rtm_read)
+
+    SlackClient = MagicMock(return_value=slack)
+    monkeypatch.setattr('layabout.SlackClient', SlackClient)
+
+    layabout.run(
+        token=TOKEN,
+        retries=1,
+        backoff=lambda r: 0,
+        until=lambda e: False
+    )
+
+
+def test_layabout_can_handle_events(monkeypatch):
+    """
+    Test that layabout can handle events appropriately given multiple event
+    handlers and even * event handlers.
+    """
+    events = [
+        dict(type='hello'),
+        dict(type='goodbye'),
+    ]
+    layabout = Layabout(name='test')
+
+    connections = [
+        True,  # Succeed with the first connection (_connect).
+    ]
+
+    rtm_connect = MagicMock(side_effect=connections)
+    rtm_read = MagicMock(return_value=events)
+    slack = MagicMock(rtm_connect=rtm_connect, rtm_read=rtm_read)
+    run_twice = MagicMock(side_effect=[True, False])
+
+    SlackClient = MagicMock(return_value=slack)
+    monkeypatch.setattr('layabout.SlackClient', SlackClient)
+
+    # Register our handlers to be called with the specified kwargs.
+    kwargs = dict(caerbannog='🐰')
+    handle_hello = MagicMock()
+    handle_goodbye = MagicMock()
+    handle_splat = MagicMock()
+    layabout.handle(type='hello', kwargs=kwargs)(fn=handle_hello)
+    layabout.handle(type='goodbye', kwargs=kwargs)(fn=handle_goodbye)
+    layabout.handle(type='*', kwargs=kwargs)(fn=handle_splat)
+
+    layabout.run(
+        token=TOKEN,
+        interval=0,
+        retries=0,
+        backoff=lambda r: 0,
+        until=run_twice
+    )
+
+    # Ensure the hello handler receives only hello events, the goodbye handler
+    # receives only goodbye events, and the * handler receives all events.
+    call_hello = call(layabout.slack, events[0], **kwargs)
+    call_goodbye = call(layabout.slack, events[1], **kwargs)
+    all_calls = [call_hello, call_goodbye]
+    handle_hello.assert_called_once_with(*call_hello[1], **call_hello[2])
+    handle_goodbye.assert_called_once_with(*call_goodbye[1], **call_goodbye[2])
+    handle_splat.assert_has_calls(all_calls)
```

Comment:
I agree and to build off that I would try to make fewer (preferably 1) assertion _per test_ and use shareable testing resources to be able to setup your fixtures/mocked objects in like 1-2 lines.

Reply:
These tests in general or just this particular one? If we're talking about `test_layabout_can_handle_events` then yeah, I could probably break it up into smaller pieces. The others generally have 1 or 2 assertions, but this one has a lot.

I think I can break the `*` handling test out into a separate test, but I felt like in order to adequately test normal event handling I needed to have more than one handler registered so I could verify they weren't accidentally called for the wrong events.

Reply:
These tests all are pretty big to me:

- `test_layabout_can_reuse_an_existing_client_to_reconnect`
- `test_layabout_can_continue_after_successful_reconnection`
- `test_layabout_can_handle_events`

Though part of that "bigness" is the testing setup duplication, which was commented on previously, and the number of newlines to keep the lines short.  On closer inspection, the tests themselves aren't doing a ton, except `can_handle_events`.

In general, I agree with @kyle-rader's advice to try and keep the tests to one assertion per test.  Although, I would expect to see at least one comprehensive test like this that has more assertions, I would expect it to be accompanied by smaller tests that break out each individual piece of functionality.  For instance, if I had test driven this event handling structure, I would have done it something like this:

- register one handler for a specific event, assert it's called
- register one handler for a specific event, assert that it _isn't_ called when a different event  is triggered
- register two handlers for the same event, trigger one, assert both are called
- register two handlers for different events, trigger one, assert the correct one is called
- register a catchall handler, trigger an event, assert it's called
- register a catchall handler, trigger several different events, assert it's called the correct number of times
- register a specific handler and a catchall, trigger the registered event, assert both are called

Now, that's quite a few tests for event triggering.  In order to make writing and reading them more palatable, you'd want to minimize the size of the setup (ideally 1 line).  This probably means having a fixture function for creating the mock slack client like we discussed above, and not using the decorated `def` style registration.  It would be fine to have a group of handler functions defined at the top level, and then register the needed ones explicitly as needed in each test.

Finally, I think it would be helpful to have a concise idiom for the assertions that lets you make each check in one line as well.  One that pops to mind is having the handlers side-effect modify a list, inserting the handler name into it. Then each assertion is pretty self-documenting, and if it fails, the problem should be immediately obvious.